Skip watching bazelignore'd paths on Linux/Windows

sushain97 opened the pull request.

sushain97 commented:
@Pavank1992 @meteorcloudy 👋 Any chance I can get some eyes on this for an initial review?

meteorcloudy commented:
@meisterT Will the performance team triage this or should the core team take a look?

joeleba commented on src/main/java/com/google/devtools/build/lib/skyframe/LocalDiffAwareness.java:
```diff
@@ -85,7 +87,7 @@ public Factory(ImmutableList<String> excludedNetworkFileSystemsPrefixes) {
     @Override
     @Nullable
-    public DiffAwareness maybeCreate(Root pathEntry) {
+    public DiffAwareness maybeCreate(Root pathEntry, ImmutableSet<com.google.devtools.build.lib.vfs.Path> ignorePaths) {
```
nit: import.

sushain97 replied:
Happy to tweak but I'm not exactly sure what change to make here! Existing code like L91 references this as `com.google.devtools.build.lib.vfs.Path` since `java.nio.file.Path` is already imported.

joeleba replied:
Ah, I missed that. Please ignore the previous comment then.

joeleba commented on src/main/java/com/google/devtools/build/lib/skyframe/LocalDiffAwareness.java:
```diff
@@ -105,7 +107,10 @@ public DiffAwareness maybeCreate(Root pathEntry) {
-      return new WatchServiceDiffAwareness(resolvedPathEntryFragment.toString());
+      return new WatchServiceDiffAwareness(
+        resolvedPathEntryFragment.toString(),
+        ignorePaths.stream().map(p -> p.toString()).collect(toImmutableSet())
```
Why do we convert `ImmutableSet<Path>` to `ImmutableSet<String>` here, and then for each of the application of `ignorePaths` in `WatchServiceDiffAwareness`, we convert it to `ImmutableSet<Path>` again? I think we should just store `ignorePaths` as `ImmutableSet<Path>`.

sushain97 replied:
This seems reasonable. I think there will still be one layer of conversion from `com.google.devtools.build.lib.vfs.Path` to `java.nio.file.Path` because `WatchServiceDiffAwareness` and related files don't use `com.google.devtools.build.lib.vfs` at all and it feels like a violation of the layering to have them use it.

joeleba commented on src/test/java/com/google/devtools/build/lib/pkgcache/IncrementalLoadingTest.java:
```diff
@@ -412,7 +413,7 @@ public void close() {
       @Nullable
       @Override
-      public DiffAwareness maybeCreate(Root pathEntry) {
+      public DiffAwareness maybeCreate(Root pathEntry, ImmutableSet<Path> ignorePaths) {
```
nit: `s/ignorePaths/ignoredPaths`. Same for the rest of the occurrences.

sushain97 replied:
No problem! I had a reason for `ignorePaths` (even though I didn't like it) but I can't find it anymore. Oh well!

joeleba commented on src/main/java/com/google/devtools/build/lib/skyframe/SkyframeExecutor.java:
```diff
@@ -3471,9 +3473,28 @@ protected WorkspaceInfoFromDiff handleDiffs(
+
+    Path workspacePath = directories.getWorkspace();
+    EvaluationResult<SkyValue> evaluationResult =
+      evaluateSkyKeys(eventHandler, ImmutableList.of(IgnoredPackagePrefixesValue.key()), false);
```
nit: Perhaps you can make `SkyframeExecutor#evaluateSkyKeys( eventHandler, skyKeys)` package-private and use that instead.

sushain97 replied:
Er, I don't think I need to change the visibility? Just dropping the third param works fine. Thanks for catching.

joeleba reviewed (changes requested):
Thanks for submitting this PR! Could you also add a test in `LocalDiffAwarenessIntegrationTest.java`?

sushain97 commented:
@joeleba I've added a test to `LocalDiffAwarenessIntegrationTest.java` and addressed your comments. That being said, I'm not super happy with the test and am open to suggestions.

I've also addressed the remaining TODO that I left.

sushain97 commented on src/main/java/com/google/devtools/build/lib/skyframe/DiffAwarenessManager.java:
```diff
@@ -38,7 +38,7 @@ public final class DiffAwarenessManager {
   // constructor; this is critical in the case where a factory always succeeds.
   private final ImmutableList<? extends DiffAwareness.Factory> diffAwarenessFactories;
-  private Map<Root, DiffAwarenessState> currentDiffAwarenessStates = Maps.newHashMap();
+  private Map<Root, Map<ImmutableSet<Path>, DiffAwarenessState>> currentDiffAwarenessStates = Maps.newHashMap();
```
> then perhaps it's worthwhile to have a static class here (maybe with @AutoValue) that encapsulates the root and the ignoredPaths and acts as the key to this mapping.

Good idea. I was looking for an easy way to do this but didn't see `@AutoValue` (I haven't written Java in many years). Thanks for the flag.

sushain97 commented:
@joeleba thanks for the suggestions. I've adjusted the key to use a simple `@AutoValue` wrapper and also added a test to verify that the `DiffAwarenessManager` correctly manages separate state for each `ignoredPaths` value.

joeleba commented:
Update: this is causing some internal tests to fail. I may need some time to look into them. Sorry for the delay!

sushain97 commented:
@joeleba gotcha. thanks for the quick update. happy to help in any way I can!

joeleba commented:
I'm working on open sourcing the failing tests, and many more.

sushain97 commented:
Sounds good! Looking forward to it.

joeleba commented:
Here it is: https://github.com/bazelbuild/bazel/commit/8304102a7d9b76a35a80e8cd2f0670f463536760. If you rebase now, you should be able to see `QueryTest#testIgnoredPackagePrefixes` failing.

sushain97 commented:
It looks like there are two failing tests after a rebase:
1. `AbstractQueryKeepGoingTest#testIgnoredPackagePrefixIsTBDQuery`
2. `QueryTest#testIgnoredPackagePrefixes`

sushain97 commented:
@joeleba Apologies for the long delay on this one! My latest commit (https://github.com/bazelbuild/bazel/pull/18363/commits/b737aa0841d3921ee69dbdb3b2df4439befc0e00) results in a green build. I believe that the issue here wasn't this change as much as the test harness itself being a bit less robust than ideal.

In particular, the tests used to trigger an evaluation of `IgnoredPackagePrefixesFunction` **after** the test writes the ignore files contents. This means that `IgnoredPackagePrefixesFunction` would return the correct non-empty set of prefixes.

However, with this change, `IgnoredPackagePrefixesFunction` is evaluated far earlier (in `SkyframeExecutor#handleDiffs`). In fact, it now happens in test setup even **before** the test itself starts writing files. This initial evaluation, before the ignore file is written, results in `IgnoredPackagePrefixesFunction` being 'cached' to the empty array and later writes don't update it since the skyframe executor never checks changed files. This means that we never invalidate the corresponding `FileValueFunction` and thus `IgnoredPackagePrefixesFunction` is never invalidated.

So, adding a call to `SkyframeExecutor#handleDiffsForTesting` before evaluating queries fixes the issue by forcing the executor to notice file writes. In practice, this is what happens in actual Bazel invocations and other tests use the same mechanism so I believe this is safe.

joeleba commented on src/test/java/com/google/devtools/build/lib/query2/testutil/SkyframeQueryHelper.java:
```diff
@@ -245,7 +245,8 @@ public AbstractBlazeQueryEnvironment<Target> getQueryEnvironment() {
   public ResultAndTargets<Target> evaluateQuery(String query)
-      throws QueryException, InterruptedException {
+      throws QueryException, InterruptedException, AbruptExitException {
+    skyframeExecutor.handleDiffsForTesting(getReporter());
```
I was thinking whether it's better to do it here, or just do it locally at the tests that concerns the ignored paths. Doing it here is probably fine since it better reflects the behavior of actual Bazel.

The reason for having this here isn't immediately obvious though, so please add a quick comment explaining this. Thanks.

joeleba reviewed (approved):
LGTM. Just the one comment left.

sushain97 commented:
@joeleba comment added! thanks for the quick review

joeleba commented:
More failing internal tests. I'm taking a look!

joeleba commented:
FYI I made some changes to fix the internal tests. Lmk if you disagree with any of them.

sushain97 commented:
I looked through the changes and it appears that the only issue we're hedging against is `diffAwareness` being `null`. That seems reasonable to me. Thank you for making those changes and avoiding another round of back-and-forth!
